src/read.rs

```diff
@@ -854,11 +856,65 @@ pub(crate) fn central_header_to_zip_file<R: Read + Seek>(
     let central_header_start = reader.stream_position()?;
 
     // Parse central header
-    let signature = reader.read_u32_le()?;
-    if signature != spec::CENTRAL_DIRECTORY_HEADER_SIGNATURE {
-        Err(ZipError::InvalidArchive("Invalid Central Directory header"))
-    } else {
-        central_header_to_zip_file_inner(reader, archive_offset, central_header_start)
+    let block = ZipEntryBlock::parse(reader)?;
+    central_header_to_zip_file_inner(reader, archive_offset, central_header_start, block)
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(packed)]
+pub(crate) struct ZipEntryBlock {
+    pub magic: spec::Magic,
+    pub version_made_by: u16,
+    pub version_to_extract: u16,
+    pub flags: u16,
+    pub compression_method: u16,
+    pub last_mod_time: u16,
+    pub last_mod_date: u16,
+    pub crc32: u32,
+    pub compressed_size: u32,
+    pub uncompressed_size: u32,
+    pub file_name_length: u16,
+    pub extra_field_length: u16,
+    pub file_comment_length: u16,
+    pub disk_number: u16,
+    pub internal_file_attributes: u16,
+    pub external_file_attributes: u32,
+    pub offset: u32,
+}
+
+impl Block for ZipEntryBlock {
+    fn interpret(bytes: Vec<u8>) -> ZipResult<Self> {
+        let block: &[u8] = bytes.as_ref();
+        let block: &[u8; mem::size_of::<Self>()] = block.try_into().unwrap();
+        let block: &Self = unsafe { mem::transmute(block) };
+        let mut block = *block;
+
+        block.magic = spec::Magic::from_le(block.magic);
+        if block.magic != spec::CENTRAL_DIRECTORY_HEADER_SIGNATURE {
+            return Err(ZipError::InvalidArchive("Invalid Central Directory header"));
+        }
+        block.version_made_by = u16::from_le(block.version_made_by);
+        block.version_to_extract = u16::from_le(block.version_to_extract);
+        block.flags = u16::from_le(block.flags);
+        block.compression_method = u16::from_le(block.compression_method);
+        block.last_mod_time = u16::from_le(block.last_mod_time);
+        block.last_mod_date = u16::from_le(block.last_mod_date);
+        block.crc32 = u32::from_le(block.crc32);
+        block.compressed_size = u32::from_le(block.compressed_size);
+        block.uncompressed_size = u32::from_le(block.uncompressed_size);
+        block.file_name_length = u16::from_le(block.file_name_length);
+        block.extra_field_length = u16::from_le(block.extra_field_length);
+        block.file_comment_length = u16::from_le(block.file_comment_length);
+        block.disk_number = u16::from_le(block.disk_number);
+        block.internal_file_attributes = u16::from_le(block.internal_file_attributes);
+        block.external_file_attributes = u32::from_le(block.external_file_attributes);
+        block.offset = u32::from_le(block.offset);
+
+        Ok(block)
+    }
+
+    fn encode(mut self) -> Vec<u8> {
```

Comment:
The lints are indicating that `self` should not be `mut`. I suspect you'll probably be able to achieve this even with the real implementation -- in fact, I suspect you can immutably borrow `&self`.

Reply:
From an aesthetic point of view, I personally like `fn encode(mut self) -> Box<[u8]>` because it consumes the intermediate "block" object. However, I can definitely see why and how to restructure this method to use `fn encode(&self) -> Box<[u8]`. Please let me know if you'd like me to change this to using `&self`.